### PR TITLE
Support partial extension UID fetch.

### PR DESCRIPTION
**Pull request by nhant01**
Support partial extension UID fetch.

### CODE REVIEW

File: core/src/main/java/com/yahoo/imapnio/async/request/AbstractFetchCommand.java [hunk 1]
```diff
@@ -96,37 +136,58 @@ protected AbstractFetchCommand(final boolean isUid, @Nonnull final String msgNum
...
+            bb.writeBytes(macro.name().getBytes(StandardCharsets.US_ASCII));
+        }
+
+        if (isUid && partialExtUidFetchInfo != null) {
+            bb.writeByte(ImapClientConstants.SPACE);
+            bb.writeByte(ImapClientConstants.L_PAREN);
+            bb.writeBytes(PARTIAL_EXTENSION_SP_B);
+            final StringBuilder sb;
+            if (partialExtUidFetchInfo.getRange() == PartialExtensionUidFetchInfo.Range.LAST) {
+                sb = new StringBuilder().append(ImapClientConstants.MINUS).append(partialExtUidFetchInfo.getLowestUid())
```
Comment (fansu) [inline core/src/main/java/com/yahoo/imapnio/async/request/AbstractFetchCommand.java:180]: why do you create StringBuilder? you can directly do bb.writeBytes(...) to save memory.

File: core/src/main/java/com/yahoo/imapnio/async/request/AbstractFetchCommand.java [hunk 2]
```diff
@@ -96,37 +136,58 @@ protected AbstractFetchCommand(final boolean isUid, @Nonnull final String msgNum
...
+                        .append(ImapClientConstants.COLON).append(ImapClientConstants.MINUS)
+                        .append(partialExtUidFetchInfo.getHighestUid()).append(ImapClientConstants.R_PAREN);
+            } else {
+                sb = new StringBuilder().append(partialExtUidFetchInfo.getLowestUid()).append(ImapClientConstants.COLON)
+                        .append(partialExtUidFetchInfo.getHighestUid()).append(ImapClientConstants.R_PAREN);
+            }
+            bb.writeBytes(sb.toString().getBytes(StandardCharsets.US_ASCII));
```
Comment (fansu) [inline core/src/main/java/com/yahoo/imapnio/async/request/AbstractFetchCommand.java:187]: you do not need string builder, please directly write to ByteBuf

File: core/src/main/java/com/yahoo/imapnio/async/request/AbstractFetchCommand.java [hunk 3]
```diff
@@ -96,37 +136,58 @@ protected AbstractFetchCommand(final boolean isUid, @Nonnull final String msgNum
...
+            bb.writeByte(ImapClientConstants.R_PAREN);
         } else {
-            sb.writeBytes(macro.name().getBytes(StandardCharsets.US_ASCII));
+            bb.writeBytes(macro.name().getBytes(StandardCharsets.US_ASCII));
+        }
+
+        if (isUid && partialExtUidFetchInfo != null) {
+            bb.writeByte(ImapClientConstants.SPACE);
+            bb.writeByte(ImapClientConstants.L_PAREN);
+            bb.writeBytes(PARTIAL_EXTENSION_SP_B);
```
Comment (fansu) [inline core/src/main/java/com/yahoo/imapnio/async/request/AbstractFetchCommand.java:177]: add SPACE and L_PAREN into PARTIAL_EXTENSION_SP_B as well, so no need to write these 3 lines

File: core/src/main/java/com/yahoo/imapnio/async/request/UidFetchCommand.java [hunk 4]
```diff
@@ -16,7 +18,19 @@ public class UidFetchCommand extends AbstractFetchCommand {
...
+    }
+
+    /**
+     * Initializes a {@link UidFetchCommand} with the {@link MessageNumberSet} array.
+     *
+     * @param msgsets the set of message set
+     * @param items the data items
+     * @param partialExtUidFetchInfo partial extension uid fetch info
+     */
+    public UidFetchCommand(@Nonnull final MessageNumberSet[] msgsets, @Nonnull final String items,
```
Comment (fansu) [inline core/src/main/java/com/yahoo/imapnio/async/request/UidFetchCommand.java:31]: not critical but a pattern. Let's not let user to pass in @Nullable partialExtUidFetchInfo, a nullable value makes parameter list looks confusing (required or not required? such). Change this to @Nonnull instead, and let line 21 to call super(...)

File: core/src/main/java/com/yahoo/imapnio/async/request/UidFetchCommand.java [hunk 5]
```diff
@@ -26,7 +40,19 @@ public UidFetchCommand(@Nonnull final MessageNumberSet[] msgsets, @Nonnull final
...
+    }
+
+    /**
+     * Initializes a {@link UidFetchCommand} with the {@link MessageNumberSet} array.
+     *
+     * @param msgsets the set of message set
+     * @param macro the macro, for example, ALL
+     * @param partialExtUidFetchInfo partial extension uid fetch info
+     */
+    public UidFetchCommand(@Nonnull final MessageNumberSet[] msgsets, @Nonnull final FetchMacro macro,
```
Comment (fansu) [inline core/src/main/java/com/yahoo/imapnio/async/request/UidFetchCommand.java:53]: not critical but a pattern. Let's not let user to pass in @Nullable partialExtUidFetchInfo, a nullable value makes parameter list looks confusing (required or not required? such). Change this to @Nonnull instead, and let line 21 to call super(...)

File: core/src/main/java/com/yahoo/imapnio/async/request/UidFetchCommand.java [hunk 6]
```diff
@@ -36,7 +62,19 @@ public UidFetchCommand(@Nonnull final MessageNumberSet[] msgsets, @Nonnull final
...
+    }
+
+    /**
+     * Initializes a {@link UidFetchCommand} with string form uids and data items.
+     *
+     * @param uids the UID string following the RFC3501 syntax. For ex:3857529045,3857529047:3857529065
+     * @param items the data items
+     * @param partialExtUidFetchInfo partial extension uid fetch info
+     */
+    public UidFetchCommand(@Nonnull final String uids, @Nonnull final String items,
```
Comment (fansu) [inline core/src/main/java/com/yahoo/imapnio/async/request/UidFetchCommand.java:75]: not critical but a pattern. Let's not let user to pass in @Nullable partialExtUidFetchInfo, a nullable value makes parameter list looks confusing (required or not required? such). Change this to @Nonnull instead, and let line 65 to call super(...)

File: core/src/main/java/com/yahoo/imapnio/async/request/UidFetchCommand.java [hunk 7]
```diff
@@ -46,7 +84,19 @@ public UidFetchCommand(@Nonnull final String uids, @Nonnull final String items)
...
+    }
+
+    /**
+     * Initializes a {@link UidFetchCommand} with string form uids and macro.
+     *
+     * @param uids the UID string following the RFC3501 syntax. For ex:3857529045,3857529047:3857529065
+     * @param macro the macro, for example, ALL
+     * @param partialExtUidFetchInfo partial extension uid fetch info
+     */
+    public UidFetchCommand(@Nonnull final String uids, @Nonnull final FetchMacro macro,
```
Comment (fansu) [inline core/src/main/java/com/yahoo/imapnio/async/request/UidFetchCommand.java:97]: not critical but a pattern. Let's not let user to pass in @Nullable partialExtUidFetchInfo, a nullable value makes parameter list looks confusing (required or not required? such). Change this to @Nonnull instead, and let line 87 to call super(...)

**Review by fansu — COMMENTED** [review]
Please check comments

**Review by fansu — APPROVED** [review]
looks good to me. @nvsnvikram please review

**Review by fansu — APPROVED** [review]

File: core/src/main/java/com/yahoo/imapnio/async/data/PartialExtensionUidFetchInfo.java [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+ *                   ; (0 < n < 4,294,967,296)
+ * }
+ * </pre>
+ */
+
+public class PartialExtensionUidFetchInfo {
+    /**
+     * Modifier instructs the server to only return FETCH results for messages in the specified range.
+     */
+    public enum Range {
```
Comment (nvsnvikram) [inline core/src/main/java/com/yahoo/imapnio/async/data/PartialExtensionUidFetchInfo.java:35]: Can we remove range?

**Review by nvsnvikram — CHANGES_REQUESTED** [review]

**Comment by fansu** [comment]
looks good to me. 
reference rfc draft: https://datatracker.ietf.org/doc/draft-ietf-extra-imap-partial/

File: core/src/main/java/com/yahoo/imapnio/async/data/PartialExtensionUidFetchInfo.java [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+ * nz-number       = digit-nz *DIGIT
+ *                   ; Non-zero unsigned 32-bit integer
+ *                   ; (0 < n < 4,294,967,296)
+ * }
+ * </pre>
+ */
+
+public class PartialExtensionUidFetchInfo {
+    /** Lowest uid number that needs to be searched. */
+    private final int lowestUid;
```
Comment (nvsnvikram) [inline core/src/main/java/com/yahoo/imapnio/async/data/PartialExtensionUidFetchInfo.java:31]: Fyi, range has only first & last. lowest:highest is equal to highest:lowest

**Review by nvsnvikram — APPROVED** [review]

**Pull request merged**